File: windows/Tokendial.Core/Model/Forecast.cs
namespace Tokendial.Core.Model;

public enum ForecastKind
{
    /// <summary>At this pace the window empties before it resets.</summary>
    RunsOut,
    /// <summary>At this pace the window lasts until its reset.</summary>
    LastsUntilReset
}

/// <summary>
/// Where the current pace takes a window. A least-squares slope over the recent samples of one epoch, because
/// vendors publish whole percentages and two neighbouring samples read as a flat line or a cliff.
/// </summary>
public sealed record Forecast(ForecastKind Kind, DateTimeOffset? RunsOutAt)
{
    public static readonly TimeSpan MinSpan = TimeSpan.FromMinutes(10);
    public static readonly TimeSpan Lookback = TimeSpan.FromMinutes(60);
    public static readonly TimeSpan Horizon = TimeSpan.FromHours(24);
    /// <summary>Half a percentage point an hour: anything slower is a resting window, not a pace.</summary>
    public const double MinSlopePerHour = 0.005;

    public static Forecast? For(IReadOnlyList<UsageSample> samples, DateTimeOffset now, DateTimeOffset? resetsAt)
    {
        var recent = samples.Where(s => now - s.TakenAt <= Lookback && s.TakenAt <= now).OrderBy(s => s.TakenAt).ToList();
        if (recent.Count < 3) return null;
        var first = recent[0];
        var last = recent[^1];
        if (last.TakenAt - first.TakenAt < MinSpan) return null;
        if (last.UsedFraction >= 1) return null;

        var slope = SlopePerHour(recent);
        if (!double.IsFinite(slope) || slope <= MinSlopePerHour) return null;
        var hoursLeft = (1 - last.UsedFraction) / slope;
        if (!double.IsFinite(hoursLeft) || hoursLeft < 0) return null;
        var runsOutAt = last.TakenAt + TimeSpan.FromHours(hoursLeft);

        var reset = resetsAt is DateTimeOffset r && r > now ? r : (DateTimeOffset?)null;
        if (reset is DateTimeOffset at) return runsOutAt < at ? new Forecast(ForecastKind.RunsOut, runsOutAt) : new Forecast(ForecastKind.LastsUntilReset, null);
        return runsOutAt - now <= Horizon ? new Forecast(ForecastKind.RunsOut, runsOutAt) : null;
    }

    /// <summary>Ordinary least squares of fraction against hours since the first sample.</summary>
    public static double SlopePerHour(IReadOnlyList<UsageSample> samples)
    {
        var origin = samples[0].TakenAt;
        var xs = samples.Select(s => (s.TakenAt - origin).TotalHours).ToArray();
        var ys = samples.Select(s => s.UsedFraction).ToArray();
        var meanX = xs.Average();
        var meanY = ys.Average();
        var covariance = xs.Zip(ys, (x, y) => (x - meanX) * (y - meanY)).Sum();
        var variance = xs.Sum(x => (x - meanX) * (x - meanX));
        return variance == 0 ? double.NaN : covariance / variance;
    }
}

File: macos/TokendialCore/Sources/TokendialCore/Model/Forecast.swift
import Foundation

public enum ForecastKind: Sendable {
    /// At this pace the window empties before it resets.
    case runsOut
    /// At this pace the window lasts until its reset.
    case lastsUntilReset
}

/// Where the current pace takes a window. A least-squares slope over the recent samples of one epoch,
/// because vendors publish whole percentages and two neighbouring samples read as a flat line or a cliff.
public struct Forecast: Equatable, Sendable {
    public var kind: ForecastKind
    public var runsOutAt: Date?

    public init(kind: ForecastKind, runsOutAt: Date?) {
        self.kind = kind
        self.runsOutAt = runsOutAt
    }

    public static let minSpan: TimeInterval = 10 * 60
    public static let lookback: TimeInterval = 60 * 60
    public static let horizon: TimeInterval = 24 * 60 * 60
    /// Half a percentage point an hour: anything slower is a resting window, not a pace.
    public static let minSlopePerHour = 0.005

    public static func of(_ samples: [UsageSample], now: Date, resetsAt: Date?) -> Forecast? {
        let recent = samples
            .filter { now.timeIntervalSince($0.takenAt) <= lookback && $0.takenAt <= now }
            .sorted { $0.takenAt < $1.takenAt }
        guard recent.count >= 3, let first = recent.first, let last = recent.last else { return nil }
        guard last.takenAt.timeIntervalSince(first.takenAt) >= minSpan else { return nil }
        guard last.usedFraction < 1 else { return nil }

        let slope = slopePerHour(recent)
        guard slope.isFinite, slope > minSlopePerHour else { return nil }
        let hoursLeft = (1 - last.usedFraction) / slope
        guard hoursLeft.isFinite, hoursLeft >= 0 else { return nil }
        let runsOutAt = last.takenAt.addingTimeInterval(hoursLeft * 3600)

        if let reset = resetsAt, reset > now {
            return runsOutAt < reset ? Forecast(kind: .runsOut, runsOutAt: runsOutAt) : Forecast(kind: .lastsUntilReset, runsOutAt: nil)
        }
        return runsOutAt.timeIntervalSince(now) <= horizon ? Forecast(kind: .runsOut, runsOutAt: runsOutAt) : nil
    }

    /// Ordinary least squares of fraction against hours since the first sample.
    public static func slopePerHour(_ samples: [UsageSample]) -> Double {
        guard let origin = samples.first?.takenAt else { return .nan }
        let xs = samples.map { $0.takenAt.timeIntervalSince(origin) / 3600 }
        let ys = samples.map { $0.usedFraction }
        let meanX = xs.reduce(0, +) / Double(xs.count)
        let meanY = ys.reduce(0, +) / Double(ys.count)
        let covariance = zip(xs, ys).reduce(0) { $0 + ($1.0 - meanX) * ($1.1 - meanY) }
        let variance = xs.reduce(0) { $0 + ($1 - meanX) * ($1 - meanX) }
        return variance == 0 ? .nan : covariance / variance
    }
}
